server/routes/createGoal/createGoalsController.ts

```diff
@@ -1,188 +1,134 @@
 import toCreateGoalDtos from '../../data/mappers/createGoalDtoMapper'
 import EducationAndWorkPlanService from '../../services/educationAndWorkPlanService'
+import {
+  CreateGoalFormAction,
+  CreateGoalsForm,
+  GoalCompleteDateOptions,
+  PartialCreateGoalsForm,
+} from './validators/GoalForm'
 
 export default class CreateGoalsController {
   constructor(private readonly educationAndWorkPlanService: EducationAndWorkPlanService) {}
 
-  getCreateGoalsView: RequestHandler = async (req, res, next): Promise<void> => {
-    const { prisonNumber } = req.params
+  getCreateGoalsView: RequestHandler = async (req, res): Promise<void> => {
     const { prisonerSummary } = req.session
 
-    const createGoalsForm = req.session.createGoalsForm || {
-      prisonNumber,
-      goals: [emptyGoal()],
-    }
-    req.session.createGoalsForm = undefined
+    const view = new CreateGoalsView(prisonerSummary, GoalCompleteDateOptions)
+    return res.render('pages/createGoals/index', { ...view.renderArgs })
+  }
 
-    const today = startOfToday()
-    const futureGoalTargetDates = [
-      futureGoalTargetDateCalculator(today, 3),
-      futureGoalTargetDateCalculator(today, 6),
-      futureGoalTargetDateCalculator(today, 12),
-    ]
+  submitAction: RequestHandler = async (req, res) => {
+    const { action, prisonNumber } = req.params
+    const goalNumber = +req.query.goalNumber
+    const stepNumber = +req.query.stepNumber
+    const createGoalsForm = req.body as PartialCreateGoalsForm
+
+    // Handle any non-submit form actions (eg: removal or addition of steps or goals)
+    let fieldID = ''
+    let updatedForm = createGoalsForm
+    if (action === CreateGoalFormAction.REMOVE_STEP) {
+      const result = handleRemoveStep(createGoalsForm, goalNumber, stepNumber)
+      if (result) {
+        updatedForm = result
+        fieldID = `#goals-${goalNumber}-steps-${updatedForm.goals[goalNumber].steps.length - 1}-title`
+      }
+    } else if (action === CreateGoalFormAction.REMOVE_GOAL) {
+      const result = handleRemoveGoal(createGoalsForm, goalNumber)
+      if (result) {
+        updatedForm = result
+        fieldID = `#goals-${goalNumber}-title`
+      }
+    } else if (action === CreateGoalFormAction.ADD_STEP) {
+      const result = handleAddAnotherStep(createGoalsForm, goalNumber)
+      if (result) {
+        updatedForm = result
+        fieldID = `#goals-${goalNumber}-steps-${updatedForm.goals[goalNumber].steps.length - 1}-title`
+      }
+    } else if (action === CreateGoalFormAction.ADD_GOAL) {
+      const result = handleAddAnotherGoal(createGoalsForm)
+      if (result) {
+        updatedForm = result
+        fieldID = `#goals-${updatedForm.goals.length - 1}-title`
+      }
+    }
 
-    const view = new CreateGoalsView(prisonerSummary, createGoalsForm, futureGoalTargetDates, req.flash('errors'))
-    return res.render('pages/createGoals/index', { ...view.renderArgs })
+    req.flash('formValues', JSON.stringify(updatedForm))
```

Comment:
Using `req.flash` for the form values might be a problem. Try this:
* Start creating a new goal for a prisoner
* Enter a description for goal 1
* Choose a date for goal 1
* Enter a step title for goal 1
* Click to add another goal
* Observe that all the field values for Goal 1 are still there and that we have an empty goal 2 👍 
* Refresh the page with Command-R
* Observe that goal 2 has gone and all the field values for goal 1 are also gone 👎 

Maybe it's not a problem? Maybe people don't refresh the page and/or accept that if they do then their field values will be gone?
But it's difference behaviour to what we currently have, and I honestly don't know if thats a problem or not? (I wonder also if there's some GDS guidance re expectations of what should happen when the user refreshes a page in this way?)